experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts

```diff
@@ -4,186 +4,163 @@
-		switch (type) {
-			case "MoveIn": {
-				const ranges = idAllocator.allocate(mark.revision ?? revision, mark.id, mark.count);
-				return ranges.map(({ first, count }) => ({
-					type: Delta.MarkType.MoveIn,
-					moveId: brandOpaque<Delta.MoveId>(first),
-					count,
-				}));
+			deltaMark.fields = deltaFromChild(changes);
+			if (deltaMark.fields.size === 0) {
+				delete deltaMark.fields;
 			}
-			case NoopMarkType: {
-				return [mark.count];
-			}
-			case "Delete": {
-				const major = mark.revision ?? revision;
-				const detachId: Delta.DetachedNodeId = { minor: mark.id };
-				if (major !== undefined) {
-					detachId.major = major;
-				}
-				return [
-					{
-						type: Delta.MarkType.Remove,
+		}
+		if (!areInputCellsEmpty(mark) && !areOutputCellsEmpty(mark)) {
+			// Since each cell is associated with exactly one node,
+			// the cell starting end ending populated means the cell content has not changed.
+			attached.push(deltaMark);
+		} else if (areInputCellsEmpty(mark) && areOutputCellsEmpty(mark)) {
+			// The cell starting and ending empty means the cell content has not changed,
+			// unless transient content was inserted/attached.
+			if (markIsTransient(mark)) {
+				const oldId = nodeIdFromChangeAtom(mark.cellId);
+				if (!areEqualChangeAtomIds(mark.cellId, mark.transientDetach)) {
+					// TODO: handle transient move-in/return-to
+					assert(isInsert(mark), "Expected non-insert transient mark");
+					if (mark.content !== undefined) {
+						build.push({
+							id: oldId,
+							trees: mark.content.map(singleTextCursor),
+						});
+					}
+					rename.push({
 						count: mark.count,
-						detachId,
-					},
-				];
-			}
-			case "MoveOut":
-			case "ReturnFrom": {
-				const ranges = idAllocator.allocate(mark.revision ?? revision, mark.id, mark.count);
-				return ranges.map(({ first, count }) => ({
-					type: Delta.MarkType.MoveOut,
-					moveId: brandOpaque<Delta.MoveId>(first),
-					count,
-				}));
+						oldId,
+						newId: nodeIdFromChangeAtom(mark.transientDetach),
+					});
+				}
+				if (deltaMark.fields) {
+					detached.push({
+						id: oldId,
+						fields: deltaMark.fields,
+					});
+				}
 			}
-			case "Insert": {
-				if (isNewAttach(mark)) {
-					assert(mark.cellId !== undefined, "Active insert mark must have CellId");
-					assert(mark.content !== undefined, "New insert must have content");
-					const cursors = mark.content.map(singleTextCursor);
-					const buildId: Delta.DetachedNodeId = {
-						minor: mark.cellId.localId,
-					};
-					if (mark.cellId.revision !== undefined) {
-						buildId.major = mark.cellId.revision;
+		} else {
+			const type = mark.type;
+			// Inline into `switch(mark.type)` once we upgrade to TS 4.7
+			switch (type) {
+				case "MoveIn": {
+					attached.push({
+						attach: makeDetachedNodeId(mark.revision ?? revision, mark.id),
+						count: mark.count,
+					});
+					break;
+				}
+				case NoopMarkType: {
```

Comment:
I think this case should never be hit as we've already checked that `mark` changes the cell content.